File: sdk/cpp/examples/basic.cpp
// LX C++ SDK - Basic Usage Example
// Copyright (c) 2025 Lux Partners Limited
// SPDX-License-Identifier: MIT

#include <lxdex/client.hpp>
#include <iostream>
#include <iomanip>
#include <thread>
#include <csignal>
#include <atomic>

std::atomic<bool> g_running{true};

void signal_handler(int signal) {
    std::cout << "\nReceived signal " << signal << ", shutting down...\n";
    g_running = false;
}

void print_orderbook(const lxdex::OrderBook& ob) {
    std::cout << "\n=== " << ob.symbol << " Orderbook ===\n";
    std::cout << std::fixed << std::setprecision(2);

    std::cout << "Asks:\n";
    size_t ask_count = std::min(ob.asks.size(), size_t{5});
    for (size_t i = ask_count; i > 0; --i) {
        const auto& level = ob.asks[i - 1];
        std::cout << "  " << std::setw(12) << level.price
                  << " | " << std::setw(10) << level.size << "\n";
    }

    std::cout << "----------------------------\n";

    std::cout << "Bids:\n";
    for (size_t i = 0; i < std::min(ob.bids.size(), size_t{5}); ++i) {
        const auto& level = ob.bids[i];
        std::cout << "  " << std::setw(12) << level.price
                  << " | " << std::setw(10) << level.size << "\n";
    }

    std::cout << "\nSpread: " << ob.spread() << " ("
              << std::setprecision(4) << ob.spread_percentage() << "%)\n";
    std::cout << "Mid: " << std::setprecision(2) << ob.mid_price() << "\n";
}

void print_trade(const lxdex::Trade& trade) {
    std::cout << "[Trade] " << trade.symbol
              << " " << (trade.side == lxdex::Side::Buy ? "BUY" : "SELL")
              << " " << trade.size << " @ " << trade.price
              << " (id: " << trade.trade_id << ")\n";
}

void print_order(const lxdex::Order& order) {
    std::cout << "[Order] " << order.symbol
              << " " << (order.side == lxdex::Side::Buy ? "BUY" : "SELL")
              << " " << order.size << " @ " << order.price
              << " status: ";

    switch (order.status) {
        case lxdex::OrderStatus::Open: std::cout << "OPEN"; break;
        case lxdex::OrderStatus::Partial: std::cout << "PARTIAL"; break;
        case lxdex::OrderStatus::Filled: std::cout << "FILLED"; break;
        case lxdex::OrderStatus::Cancelled: std::cout << "CANCELLED"; break;
        case lxdex::OrderStatus::Rejected: std::cout << "REJECTED"; break;
    }
    std::cout << " (id: " << order.order_id << ")\n";
}

int main(int argc, char* argv[]) {
    // Setup signal handler
    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);

    std::cout << "LX C++ SDK Example\n";
    std::cout << "======================\n\n";

    // Configure client
    lxdex::ClientConfig config;
    config.ws_url = "ws://localhost:8081";
    config.auto_reconnect = true;
    config.max_reconnect_attempts = 5;

    // Parse command line args
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--url" && i + 1 < argc) {
            config.ws_url = argv[++i];
        } else if (arg == "--key" && i + 1 < argc) {
            config.api_key = argv[++i];
        } else if (arg == "--secret" && i + 1 < argc) {
            config.api_secret = argv[++i];
        } else if (arg == "--help") {
            std::cout << "Usage: " << argv[0] << " [options]\n"
                      << "  --url URL      WebSocket URL (default: ws://localhost:8081)\n"
                      << "  --key KEY      API key for authentication\n"
                      << "  --secret SEC   API secret for authentication\n"
                      << "  --help         Show this help\n";
            return 0;
        }
    }

    // Create client
    auto client = lxdex::make_client(config);

    // Setup callbacks
    client->on_error([](const lxdex::Error& err) {
        std::cerr << "[Error] " << err.message;
        if (!err.request_id.empty()) {
            std::cerr << " (request: " << err.request_id << ")";
        }
        std::cerr << "\n";
    });

    client->on_connection([](lxdex::ConnectionState state) {
        std::cout << "[Connection] State: ";
        switch (state) {
            case lxdex::ConnectionState::Disconnected:
                std::cout << "DISCONNECTED";
                break;
            case lxdex::ConnectionState::Connecting:
                std::cout << "CONNECTING";
                break;
            case lxdex::ConnectionState::Connected:
                std::cout << "CONNECTED";
                break;
            case lxdex::ConnectionState::Reconnecting:
                std::cout << "RECONNECTING";
                break;
            case lxdex::ConnectionState::Failed:
                std::cout << "FAILED";
                break;
        }
        std::cout << "\n";
    });

    client->on_order([](const lxdex::Order& order) {
        print_order(order);
    });

    // Connect
    std::cout << "Connecting to " << config.ws_url << "...\n";
    auto err = client->connect();
    if (err) {
        std::cerr << "Failed to connect: " << err.message << "\n";
        return 1;
    }
    std::cout << "Connected!\n\n";

    // Ping test
    auto ping_result = client->ping();
    if (ping_result.ok()) {
        std::cout << "Ping: " << ping_result.value << " us\n\n";
    }

    // Authenticate if credentials provided
    if (!config.api_key.empty()) {
        std::cout << "Authenticating...\n";
        err = client->authenticate();
        if (err) {
            std::cerr << "Authentication failed: " << err.message << "\n";
        } else {
            std::cout << "Authenticated!\n\n";
        }
    }

    // Subscribe to orderbook updates
    std::vector<std::string> symbols = {"BTC-USDT", "ETH-USDT"};

    std::cout << "Subscribing to orderbook updates...\n";
    err = client->subscribe_orderbook(symbols, [](const lxdex::OrderBook& ob) {
        print_orderbook(ob);
    });
    if (err) {
        std::cerr << "Failed to subscribe: " << err.message << "\n";
    }

    // Subscribe to trades
    std::cout << "Subscribing to trade updates...\n";
    err = client->subscribe_trades(symbols, [](const lxdex::Trade& trade) {
        print_trade(trade);
    });
    if (err) {
        std::cerr << "Failed to subscribe: " << err.message << "\n";
    }

    // Example: Place a test order (if authenticated)
    if (client->is_authenticated()) {
        std::cout << "\nPlacing test order...\n";

        lxdex::Order order;
        order.symbol = "BTC-USDT";
        order.side = lxdex::Side::Buy;
        order.type = lxdex::OrderType::Limit;
        order.price = 40000.0;
        order.size = 0.001;
        order.time_in_force = lxdex::TimeInForce::GTC;
        order.client_id = lxdex::Client::generate_client_id();

        auto result = client->place_order(order);
        if (result.ok()) {
            std::cout << "Order placed! ID: " << result.value.order_id
                      << " Status: " << result.value.status << "\n";

            // Cancel the order after a brief delay
            std::this_thread::sleep_for(std::chrono::seconds(2));

            std::cout << "Cancelling order...\n";
            err = client->cancel_order(result.value.order_id);
            if (err) {
                std::cerr << "Failed to cancel: " << err.message << "\n";
            } else {
                std::cout << "Order cancelled!\n";
            }
        } else {
            std::cerr << "Failed to place order: " << result.error.message << "\n";
        }
    }

    // Main loop - wait for data
    std::cout << "\nWaiting for market data (Ctrl+C to exit)...\n\n";

    int update_count = 0;
    while (g_running) {
        std::this_thread::sleep_for(std::chrono::seconds(1));

        // Periodic metrics report
        if (++update_count % 30 == 0) {
            auto metrics = client->metrics();
            std::cout << "\n--- Metrics ---\n"
                      << "Messages sent:     " << metrics.messages_sent << "\n"
                      << "Messages received: " << metrics.messages_received << "\n"
                      << "Orders sent:       " << metrics.orders_sent << "\n"
                      << "Trades received:   " << metrics.trades_received << "\n"
                      << "Last latency:      " << metrics.last_latency_us << " us\n"
                      << "Reconnects:        " << metrics.reconnect_count << "\n"
                      << "Errors:            " << metrics.error_count << "\n\n";
        }

        // Check connection
        if (!client->is_connected()) {
            std::cout << "Connection lost, waiting for reconnect...\n";
        }
    }

    // Cleanup
    std::cout << "\nDisconnecting...\n";
    client->disconnect();
    std::cout << "Done.\n";

    return 0;
}

File: sdk/rust/examples/basic.rs
//! Basic trading example for LX DEX.
//!
//! Demonstrates:
//! - Connecting to WebSocket
//! - Authentication
//! - Subscribing to market data
//! - Placing and canceling orders
//! - Handling events

use lux_dex::{Client, ClientConfig, Order, Side, WsEvent};
use std::env;
use tracing::{info, warn};

#[tokio::main]
async fn main() -> lux_dex::Result<()> {
    // Initialize logging
    tracing_subscriber::fmt::init();

    // Configuration from environment or defaults
    let ws_url = env::var("LX_WS_URL").unwrap_or_else(|_| "ws://localhost:8081".into());
    let http_url = env::var("LX_HTTP_URL").unwrap_or_else(|_| "http://localhost:8080".into());
    let api_key = env::var("LX_API_KEY").ok();
    let api_secret = env::var("LX_API_SECRET").ok();

    let mut config = ClientConfig::default()
        .with_ws_url(&ws_url)
        .with_http_url(&http_url);

    if let (Some(key), Some(secret)) = (api_key, api_secret) {
        config = config.with_credentials(key, secret);
    }

    // Create client
    let client = Client::with_config(config);

    // Take event receiver before connecting
    let mut events = client.take_event_receiver().await.unwrap();

    // Connect to WebSocket
    info!("Connecting to {}", ws_url);
    client.connect().await?;

    // Spawn event handler
    let event_handle = tokio::spawn(async move {
        while let Some(event) = events.recv().await {
            match event {
                WsEvent::Connected { client_id } => {
                    info!("Connected with client_id: {}", client_id);
                }
                WsEvent::Authenticated { user_id } => {
                    info!("Authenticated as user: {}", user_id);
                }
                WsEvent::OrderBook(book) => {
                    info!(
                        "Order book {}: bid={:?} ask={:?} spread={:?}",
                        book.symbol,
                        book.best_bid(),
                        book.best_ask(),
                        book.spread()
                    );
                }
                WsEvent::Trade(trade) => {
                    info!(
                        "Trade: {} {} @ {} (value: {})",
                        trade.symbol,
                        trade.size,
                        trade.price,
                        trade.value()
                    );
                }
                WsEvent::OrderUpdate { order, status } => {
                    info!(
                        "Order update: {} {} {} @ {} - {}",
                        order.order_id.unwrap_or(0),
                        order.side,
                        order.size,
                        order.price,
                        status
                    );
                }
                WsEvent::Position(pos) => {
                    info!(
                        "Position: {} {} size={} entry={} pnl={}",
                        pos.symbol, pos.side, pos.size, pos.entry_price, pos.unrealized_pnl
                    );
                }
                WsEvent::Price { symbol, price } => {
                    info!("Price update: {} = {}", symbol, price);
                }
                WsEvent::Error { message, request_id } => {
                    warn!("Error: {} (request: {:?})", message, request_id);
                }
                WsEvent::Pong => {
                    info!("Pong received");
                }
                _ => {}
            }
        }
    });

    // Wait for connection
    tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;

    // Authenticate if credentials are configured
    if client.is_connected() {
        info!("Attempting authentication...");
        client.authenticate_configured().await.ok();
    }

    // Subscribe to market data
    info!("Subscribing to BTC-USDT order book");
    client.subscribe_orderbook("BTC-USDT").await?;

    info!("Subscribing to ETH-USDT trades");
    client.subscribe_trades("ETH-USDT").await?;

    // Get order book via HTTP
    info!("Fetching order book via HTTP...");
    match client.get_orderbook("BTC-USDT", 10).await {
        Ok(book) => {
            info!(
                "HTTP Order book: {} levels, mid={:?}",
                book.bids.len() + book.asks.len(),
                book.mid_price()
            );

            // Demonstrate slippage estimation
            if let Some(avg_price) = book.estimate_buy_slippage(1.0) {
                info!("Estimated buy price for 1 BTC: {}", avg_price);
            }
        }
        Err(e) => {
            warn!("Failed to fetch order book: {}", e);
        }
    }

    // Place a test order (only if authenticated)
    tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;

    if client.is_authenticated() {
        info!("Placing test limit order...");
        let order = Order::limit("BTC-USDT", Side::Buy, 45000.0, 0.001)
            .with_client_id("rust-sdk-test-001")
            .post_only();

        client.place_order(&order).await?;
        info!("Order placed");

        // Wait for confirmation
        tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
    }

    // Ping to keep connection alive
    client.ping().await?;

    // Run for a while to receive events
    info!("Listening for events... (press Ctrl+C to exit)");
    tokio::select! {
        _ = tokio::signal::ctrl_c() => {
            info!("Shutting down...");
        }
        _ = event_handle => {
            info!("Event handler finished");
        }
    }

    // Disconnect
    client.disconnect().await?;
    info!("Disconnected");

    Ok(())
}
